lib/render/RayCaster.cpp

```diff
@@ -430,41 +480,88 @@ bool RayCaster::UserCoordinates::UpdateCoordinates( const RayCasterParams* param
     {
         for( size_t i = 0; i < numOfVertices; i++ )
         {
-            dataField[ i ] = (float(*valItr) - valueRange[0]) * valueRange1o;
+            dataField[ i ] = float(*valItr);
             ++valItr;
         }
     }
 
     delete grid;
-    return true;
+    return 0;
 }
 
-int RayCaster::_initializeGL()
+int  RayCaster::UserCoordinates::UpdateCurviCoords( const RayCasterParams* params,
+                                                          DataMgr*         dataMgr )
 {
-#ifdef Darwin
-    return 0;
-#endif
-    // Enable debug output
-    glEnable              ( GL_DEBUG_OUTPUT );
-    glDebugMessageCallback( MessageCallback, 0 );
-
-    _loadShaders();
+    assert( params->GetCastingMode() == 2 );
+
+    if( xyCoords )
+        delete[] xyCoords;
+    xyCoords = new float[ dims[0] * dims[1] * 2 ];
+    if( zCoords )
+        delete[] zCoords;
+    zCoords  = new float[ dims[0] * dims[1] * dims[2] ];
+    if( !zCoords )  // Test if allocation successful for 3D buffers. 
+    {
+        MyBase::SetErrMsg("Failed to allocate memory");
+        return -1;
+    }
 
-    // Create Vertex Array Object (VAO)
-    glGenVertexArrays( 1, &_vertexArrayId );
-    glGenBuffers(      1, &_vertexBufferId );
-    glGenBuffers(      1, &_indexBufferId );
+    // Gather the XY coordinate from frontFace buffer
+    size_t xyIdx = 0, xyzIdx = 0;
+    for( size_t y = 0; y < dims[1]; y++ )
+        for( size_t x = 0; x < dims[0]; x++ )
+        {
+            xyCoords[ xyIdx++ ] = frontFace[xyzIdx++];
+            xyCoords[ xyIdx++ ] = frontFace[xyzIdx++];
+            xyzIdx++;
+        }
 
-    _initializeFramebufferTextures();
+    // Gather the Z coordinates from grid
+    StructuredGrid* grid = nullptr;
+    if( this->GetCurrentGrid( params, dataMgr, &grid ) != 0 )
+    {
+        MyBase::SetErrMsg("Failed to retrieve a StructuredGrid");
+        return 1;
+    }
+    StructuredGrid::ConstCoordItr coordItr   = grid->ConstCoordBegin();
+    size_t numOfVertices = dims[0] * dims[1] * dims[2];
+    for( xyzIdx = 0; xyzIdx < numOfVertices; xyzIdx++ )
+    {
+        zCoords[xyzIdx] = float((*coordItr)[2]);
+        ++coordItr;
+    }
 
     return 0;
 }
 
+int RayCaster::_initializeGL()
+{
+    if( _loadShaders() != 0 )
+    {
+        MyBase::SetErrMsg("Failed to load shaders!");
+        return -1;
+    }
+    if( _initializeFramebufferTextures() != 0 )
+    {
+        MyBase::SetErrMsg("Failed to Create Framebuffer and Textures!");
+        return -1;
+    }
+
+    return 0;   // Success
+}
+
 int RayCaster::_paintGL( bool fast ) 
```

Comment:
The paingGL function is way to long. Furthermore, it doesn't restore opengl state on an error condition. Please break it up into smaller, well-defined pieces.

Reply:
You can easily consolidate the 6 pairs of face rendering nested loops by taking advantage of the fact that very little changes between each pair (e.g. left/right, top/bottom, front/back). For example, the front and back triangle rendering loops differ by constant values. In the first loop you have:

`unsigned int   attribIdx       = ((y + 1) * bx + x) * 4;
attrib1Buffer[ attribIdx + 2 ] = int(bz) - 2;
attrib1Buffer[ attribIdx + 3 ] = 0;`

and in the second loop you have:

`unsigned int   attribIdx       = (y * bx + x) * 4;
attrib1Buffer[ attribIdx + 2 ] = 0;
attrib1Buffer[ attribIdx + 3 ] = 1;`

You could have a single function for both of these two loops parameterized by a front/back flag:

`if (frontFlag) {
yinc = 1
att2 = int(bz) - 2;
att3 = 0
} else {
yinc = 0
att2 = 0
att3 = 1
}`

Then your inner loop would look like:

`attribIdx       = ((y + yinc) * bx + x) * 4;
attrib1Buffer[ attribIdx + 2 ] = att2
attrib1Buffer[ attribIdx + 3 ] = att3
`

This would make _renderTriangleStrips() much shorter, and make the face rendering code more maintainable.